File: Next_Permutation.cpp
/*31. Next Permutation  QuestionEditorial Solution  My Submissions
Total Accepted: 83164 Total Submissions: 299955 Difficulty: Medium
Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.

If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).

The replacement must be in-place, do not allocate extra memory.

Here are some examples. Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.
1,2,3 → 1,3,2
3,2,1 → 1,2,3
1,1,5 → 1,5,1
Hide Company Tags Google
Hide Tags Array
Hide Similar Problems (M) Permutations (M) Permutations II (M) Permutation Sequence (M) Palindrome Permutation II
*/



/*O(n) time
permutation 置换
1.Find the largest index k such that nums[k] < nums[k + 1]. If no such index exists,
	the permutation is sorted in descending order, just reverse it to ascending order and we are done. For example, the next permutation of [3, 2, 1] is [1, 2, 3].
2. Find the largest index l greater than k such that nums[k] < nums[l].
3. Swap the value of nums[k] with that of nums[l].
4. Reverse the sequence from nums[k + 1] up to and including the final element
	nums[nums.size() - 1]*/

//C++
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int k = -1;
        //find the largest index k, that nums[k] < nums[k + 1]
        for(int i = nums.size()-2; i >= 0; --i){
            if(nums[i]<nums[i+1]){
                k=i;
                break;
            }
        }

        //if k==-1, the permutation is sorted in descending order
        //reverse it into ascending order
        if(k==-1){
            reverse(nums.begin(), nums.end());
            return;
        }

        //find the largest index l greater than k, nums[k]<nums[l]
        int l = -1;
        for(int i = nums.size()-1; i > k; --i){
            if(nums[k] < nums[i]){
                l = i;
                break;
            }
        }

        swap(nums[k], nums[l]);
        reverse(nums.begin()+k+1, nums.end());
    }
};

File: Next_Permutation.swift
/*
Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.
If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).
The replacement must be in-place and use only constant extra memory.
Here are some examples. Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.
1,2,3 → 1,3,2
3,2,1 → 1,2,3
1,1,5 → 1,5,1

看是否有下一个排列组合的数字 比当前数字稍微大一些
如果没有，直接逆转排序
*/

/*
Solution1: Brute Force
Just find all of the possiblity of permutation and find out the permutation which is just larger than the given one. But this one will be a very naive approach, since it requires us to find out every possible permutation which will take really long time and the implementation is complex. Thus, this approach is not acceptable at all.
//
Time complexity : O(n!). Total possible permutations is n!.
Space complexity : O(n). Since an array will be used to store the permutations.
*/

/*
Solution 2:
1. find first a[first] which is in the ascending order, if cannot find, just reverse this array would be the answer. Note: after finding this element, assign it to a variable for checking, otherwise it is hard to clarify the descending & target index at 0
2. betwen firstIndex+1 to n-1, find the element just larger than a[firstIndex], and swap them
3. now, a[firstIndex] is the fixed one, for firstIndex+1 to n-1, just reverse them would be the correct answer
//
Time complexity : O(n). In worst case, only two scans of the whole array are needed.
Space complexity : O(1). No extra space is used. In place replacements are done.
*/
class Solution {
    func nextPermutation(_ nums: inout [Int]) {
        let n = nums.count

        // find first index which is ascending order
        var index = n-2
        while index >= 0 {
            if nums[index] < nums[index+1] {
                break
            }
            index -= 1
        }

        if index == -1 {
            // nums is descending array, which is already largest
            // return lowest possible order
            nums.reverse()
            return
        }

        // find index+1..<n, the closest digit larger than nums[index]
        var next = index+1
        for i in next..<n {
            if nums[i] > nums[index], i > next {
                next = i
            }
        }

        // swap index, next
        nums.swapAt(index, next)

        // sort index+1 to inscending order
        nums[(index+1)...].reverse()
    }
}